### ISSUE TITLE
BERT Loses Patience: Fast and Robust Inference with Early Exit

### ISSUE THREAD

**Issue by ymgw55** [ymgw55/MLPapers#157]
# Information
### URL
https://proceedings.neurips.cc/paper/2020/hash/d4dd111a4fd973394238aca5c05bebe3-Abstract.html
### Authors
Wangchunshu Zhou, Canwen Xu, Tao Ge, Julian McAuley, Ke Xu, Furu Wei
### Published at
NeurIPS 2020

### Code
https://github.com/JetRunner/PABEE
# Reading Motivation
#127 で出てきた, BERT-based モデルの overthingking について学ぶ.
# Summary
Patience-based な Early Exit という手法を提案. pretrain された言語モデル (PLM) で効率的かつロバストな推論をする手法. これは, PLM の各層で推論を行い, あらかじめ決めた patient  数だけ推論結果が変化しなければ, 動的に推論を打ち切ることによって達成. 少ない層で推論するので効率があがり, 一方で言語モデルの overthinking を防ぐことで精度とrobustness が向上. また, 最終的な予測出力は, 複数の layer での予測を見て決めるので, 既存の early exit 手法とくらべ, accuracy と seepd の trade-off がより良い.

# Background
* 近年の言語モデルは層・パラメータが多く, 計算コストが高い.
* パラメータ数が多いことで, "overthinking" と呼ばれる, 最初の方の層では正しい予測ができるが, 最終層では, 関係のない特徴に予測が左右されてしまう現象がある. 
# Contribution
# Proposed Method
まず, finetuning は, 各 layer から確率を取り出せるような最終層を追加しておく. 
次にあらかじめ patient t を決める (例えば t = 6). 分類・回帰, それぞれの場合で以下の更新式カウントし, t を超えたところで予測結果とする. つまり, 分類の場合 t 回, inner layer の予測が同じ, 回帰の場合, t 回, layer 間の予測が threshold tau 以下の時, カウントをインクリメント. そうでは無ければ 0 にリセットする. 

<img width="608" alt="スクリーンショット 2022-05-11 093720" src="https://user-images.githubusercontent.com/39965001/167746197-2a0162da-8f8c-4d23-9e0d-d3c64699d3d7.png">

Figure1 では, 既存手法の Shallow-Deep Net では, 確率が threshold を超えるとそこで exit するため, 誤った出力をしているが, PABEE では, 複数の layer の出力を考慮するので正しい出力となっている.
<img width="602" alt="スクリーンショット 2022-05-11 092129" src="https://user-images.githubusercontent.com/39965001/167744996-5f397849-c2d4-48c2-9b84-44f6e5dc54ce.png">
# Experiment
ALBERT について, 層が深くなると, entropy が下がる (= confidence が上がる) 一方, Layer = 10 を境に error rate は上がる.
<img width="605" alt="スクリーンショット 2022-05-11 092743" src="https://user-images.githubusercontent.com/39965001/167745449-03e3647b-3ad5-48b7-ac94-5218de80428e.png">

baseline に PABEE を加えると, GLUE タスクでスコア向上.
<img width="604" alt="スクリーンショット 2022-05-11 092951" src="https://user-images.githubusercontent.com/39965001/167745629-ac599b46-eb83-49dc-be59-42f09a379c09.png">

# Consideration
入力文ごとに使う hidden_state が異なるので batch_size = 1 で推論する必要があり, 微妙... ただ最終 layer の出力に拘らなくてもよいという事は分かった.
